### PR TITLE
Driver e2e tests

### PR DESCRIPTION
**Pull request by crazy-max**
Follow-up #691 

This workflow handles multi-matrix jobs with:
* buildkit image (latest and master)
* buildkit cfg
* driver (docker, container and kubernetes)
* multi-node (restrict to kubernetes atm)

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

### CODE REVIEW

File: .github/workflows/e2e.yml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+          - 0
+          - 1
+        multi-node:
+          - 0
+          - 1
+        platforms:
+          - linux/amd64,linux/arm64
+        include:
+          - driver: kubernetes
+            driver-opt: qemu.install=true
```
Comment (crazy-max) [inline .github/workflows/e2e.yml:37]: @morlay `driver-opt` can be used for a specific driver like `qemu.install=true` for kubernetes.

File: .github/workflows/e2e.yml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        driver:
+          - docker
+          - docker-container
+          - kubernetes
+        buildkit:
+          - moby/buildkit:buildx-stable-1
+          - moby/buildkit:master
+        buildkit-cfg:
+          - 0
+          - 1
```
Comment (crazy-max) [inline .github/workflows/e2e.yml:29]: Support for `--config` is also tested for kubernetes and container drivers.

File: hack/test-driver [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+RUN echo "I am running on \$BUILDPLATFORM, building for \$TARGETPLATFORM" > /log
+FROM busybox
+COPY --from=build /log /log
+RUN cat /log
+RUN uname -a
+EOL
+
+# build
+${BUILDX_CMD} build ${buildPlatformFlag} \
+  --output="type=cacheonly" \
```
Comment (crazy-max) [inline hack/test-driver:97]: For now I use `type=cacheonly` as output but in the future we could push to a local registry using [GHA services](https://github.com/docker/build-push-action/blob/375f72aff65b645c51f122f04a776df8416b222b/.github/workflows/ci.yml#L153-L157) or private registries like we do in [build-push-action](https://github.com/docker/build-push-action/blob/master/.github/workflows/e2e.yml).

File: .github/workflows/e2e.yml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      - 'master'
+      - 'v[0-9]*'
+  pull_request:
+    branches:
+      - 'master'
+      - 'v[0-9]*'
+
+jobs:
+  driver:
+    runs-on: ubuntu-latest
```
Comment (AkihiroSuda) [inline .github/workflows/e2e.yml:16]: nit: ubuntu-20.04 might be slightly better for reproducibility

File: .github/workflows/e2e.yml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      fail-fast: false
+      matrix:
+        driver:
+          - docker
+          - docker-container
+          - kubernetes
+        buildkit:
+          - moby/buildkit:buildx-stable-1
+          - moby/buildkit:master
+        buildkit-cfg:
```
Comment (AkihiroSuda) [inline .github/workflows/e2e.yml:27]: better to use a form like `buildkit-cfg: ["buildkit-cfg-true", "buildkit-cfg-false"]`
(Please see the screenshot in my other comment)

Comment (crazy-max) [reply .github/workflows/e2e.yml:27]: sgtm

**Comment by crazy-max** [comment]
PTAL @AkihiroSuda @morlay

File: .github/workflows/e2e.yml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            buildkit-cfg: bkcfg-true
+          - driver: docker-container
+            multi-node: mnode-true
+    steps:
+      -
+        name: Checkout
+        uses: actions/checkout@v2
+      -
+        name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
```
Comment (morlay) [inline .github/workflows/e2e.yml:51]: This step may only when `driver` is `docker` or `docker-container`

could be

```
uses: docker/setup-qemu-action@v1
if:  ${{ matrix.driver == 'docker' && matrix.driver == 'docker-container'  }}
```

**Review by AkihiroSuda — APPROVED** [review]

**Review by tonistiigi — APPROVED** [review]
Let's merge this but as a follow-up, I'd like `./test-driver` to run inside a container. So it is not specific to github actions but can be invoked from any system.

`./hack/shell` already creates such a container today. I tried running `./test-driver` inside that container and it worked for both docker and container driver. It doesn't work for k3s atm as it is not installed but I don't see a big reason why it couldn't be.

We can also discuss in later if it makes sense to use bats or go tests for this but ok for a starter.

**Pull request merged**